Implement wheel WPT actions in WebKit's vendor JS

nmoucht opened the pull request:
#### c84e12023b1fdceb3db72041895d5e5160ad7acf
<pre>
Implement wheel WPT actions in WebKitTestRunner
<a href="https://bugs.webkit.org/show_bug.cgi?id=243272">https://bugs.webkit.org/show_bug.cgi?id=243272</a>
&lt;rdar://97691874&gt;

Reviewed by NOBODY (OOPS!).

WIP patch from Tim Nguyen. Implement wheel actions in WebKitTestRunner.

* LayoutTests/imported/w3c/web-platform-tests/css/css-overscroll-behavior/overscroll-behavior-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-snap/input/snap-area-overflow-boundary-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-snap/snap-at-user-scroll-end-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
(async dispatchWheelActions):
(async let):
(window.test_driver_internal.action_sequence):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c84e12023b1fdceb3db72041895d5e5160ad7acf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91567 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/732 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22190 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101276 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/161338 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/732 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/29427 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83891 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97624 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97225 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/452 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78252 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27404 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82376 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/82030 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70443 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35691 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16059 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33458 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17153 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37285 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/39928 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39192 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36270 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

smfr commented on LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
```diff
@@ -145,6 +145,41 @@ function dispatchTouchActions(actions, options = { insertPauseAfterPointerUp: fa
+            x += bounds.left + 1;
+            y += bounds.top + 1;
+        }
+        eventSender.monitorWheelEvents();
```
I'd prefer this use the more modern `uiController.sendEventStream()`; see `UIHelper.mouseWheelSequence()`.

nmoucht replied:
Using this causes tests to hang that passed using `mouseScrollByWithWheelAndMomentumPhases`. When using `await uiController.sendEventStream()`, I saw the events were not even getting to the UIScriptController, but without awaiting they were getting though. My assumption is that it may be necessary to call `await uiController.sendEventStream()` on another thread, which the tests are not doing. I think it makes sense to stick with the previous method for now.

smfr commented on LayoutTests/TestExpectations:
```diff
@@ -5984,3 +5984,12 @@ webkit.org/b/243562 imported/w3c/web-platform-tests/css/css-ui/compute-kind-widg
 imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/css-module/cors-crossorigin-requests.html [ Pass Failure ]
 imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/cors-crossorigin-requests.html [ Pass Failure ]
+
+# These are timeouts after adding wheel actions to WebKitTestRunner: webkit.org/b/243272
```
Describing this as adding wheel actions "to WebKitTestRunner" is misleading. It would be more appropriate to say "in WPT vendor JS".

nmoucht replied:
Fixed.

nmoucht committed:
Implement wheel WPT actions in WebKitTestRunner
https://bugs.webkit.org/show_bug.cgi?id=243272
<rdar://97691874>

Reviewed by NOBODY (OOPS!).

WIP patch from Tim Nguyen. Implement wheel actions in WebKitTestRunner.

* LayoutTests/imported/w3c/web-platform-tests/css/css-overscroll-behavior/overscroll-behavior-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-snap/input/snap-area-overflow-boundary-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-snap/snap-at-user-scroll-end-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
(async dispatchWheelActions):
(async let):
(window.test_driver_internal.action_sequence):

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-overscroll-behavior/overscroll-behavior-expected.txt:
```diff
@@ -9,7 +8,5 @@ Repeat the same scrolls as in step 1 and then tap on DONE.
-Harness Error (FAIL), message = Unhandled rejection: Can't find variable: action
-
-NOTRUN overscroll-behavior prevents scroll-propagation in the area and direction as specified
+PASS overscroll-behavior prevents scroll-propagation in the area and direction as specified
```
This doesn't match the behavior upstream that WebDriver implements: https://wpt.fyi/results/css/css-overscroll-behavior?label=master&label=experimental&product=chrome&product=firefox&product=safari&aligned&view=interop&q=label%3Ainterop-2022-scrolling%20css%2Fcss-overscroll-behavior%2Foverscroll-behavior

I think consistency is important here.

nmoucht replied:
Yeah I have tried investigating this a bit with a local WebDriver and the behavior on this test is strange. If I remove the overscroll-behavior values set it is expected that both scrolls would propagate, but that is not the behavior I was seeing (only the vertical scroll was propagating). I also noticed that for vertical wheel events we were getting multiple opposite events (positive then negative events). I'm still trying to understand where along the flow things are broken. I'm hoping to come back to this soon.
